File: utils/sort_by_date.go
package utils

import (
	"rahul-rakshit/formatbankcsv/output"
	"sort"
	"time"
)

func SortByDate(input [][]string) [][]string {
	type DateValuePair struct {
		Date  string
		Value []string
	}

	var data []DateValuePair

	for _, item := range input {
		data = append(data, DateValuePair{
			Date:  item[0],
			Value: item[1:],
		})
	}

	sort.Slice(data, func(i, j int) bool {
		if data[i].Date == output.Header[0] {
			return true
		} else if data[j].Date == output.Header[0] {
			return false
		}

		date1, err1 := time.Parse("2006-01-02", data[i].Date)
		date2, err2 := time.Parse("2006-01-02", data[j].Date)

		if err1 != nil || err2 != nil {
			return false
		}

		return date1.Before(date2)
	})

	var sorted [][]string

	for _, item := range data {
		sortedItem := append([]string{item.Date}, item.Value...)
		sorted = append(sorted, sortedItem)
	}

	return sorted
}

File: utils/sortByDate.ts
import { Header } from "./constants.ts";

interface DateValuePair {
  date: string;
  value: string[];
}

export function sortByDate(input: string[][]): string[][] {
  const data: DateValuePair[] = input.map((item) => ({
    date: item[0],
    value: item.slice(1),
  }));

  data.sort((a, b) => {
    if (a.date === Header[0]) return -1;
    if (b.date === Header[0]) return 1;

    try {
      const date1 = new Date(a.date);
      const date2 = new Date(b.date);

      if (isNaN(date1.getTime()) || isNaN(date2.getTime())) {
        return 0;
      }

      return date1.getTime() - date2.getTime();
    } catch {
      return 0;
    }
  });

  return data.map((item) => [item.date, ...item.value]);
}
